### ISSUE TITLE
Allow `vim.{o,bo,wo,go}` meta accessors to work with tables

### ISSUE THREAD

**Issue by clason** [neovim/neovim#20107]
### Feature description

###Background

The `vim.{o,bo,wo,go}` metamethods allow getting and setting Neovim options from Lua and are full replacements for the `:set` and `:setlocal` commands. (The only drawback of `vim.bo` and `vim.wo` over `:setlocal` is that users need to know whether a local option is buffer-local or window-local, but 1. that information is easily looked up via `:h '<option>'` and 2. this is more than made up for by being able to target specific buffers via `vim.bo[bufnr]`.)

However, Vim has a number of string options that are actually lists (either comma separated like `runtimepath` or concatenated characters like `formatoptions` or dictionaries (like `listchars`), which are cumbersome to manipulate as strings from Lua. For this, `vim.opt` offers a convenient special (object-oriented) table-based interface:
* `vim.opt.listchars:append({tab = '>-'})`
* `vim.opt.listchars:prepend({tab = '>-'})`

**Comment by justinmk** [comment]
Related: https://github.com/neovim/neovim/issues/18246

**Comment by binyomen** [comment]
Thanks for picking this up again! I'm really excited to see this work! I'm happy to lend my time to help design and implement this.

I'd say that there's still value in adding a `__newindex` metamethod to returned tables. These wouldn't be plain tables returned from `vim.o` in that case, but users would be able to interact with them more similarly to how they interact with plain Lua tables, i.e. by mutating (or in this case appearing to mutate) a shared reference to an underlying object. This would (at least in my opinion) produce more idiomatic Lua and enable more standard patterns of interacting with tables, since tables returned from `vim.o` would behave more closely to how tables without `__index`/`__newindex` usually behave. I guess my suggestion is basically to add metamagic to smooth over the metamagic that already has to exist in `vim.o` itself, with the former helping conceal the latter. That said I fully understand the arguments against this approach, and my suggestion would be implementable as a fairly basic wrapper API in a plugin (like adding a `vim.my_o` table that calls `dict_set_key` and friends under the hood), but just wanted to give my two cents 😄

Also, is the reason some of the suggested functions start with `dict` as opposed to `tbl` to declare them as being specifically for tables like `vim.o` and `vim.g` that require direct assignment rather than field mutation? Personally I think it still makes sense to use `tbl` for those, since they're also generally useful for table manipulation where you might want to follow a functional pattern (e.g. returning a copy of the original table but with a different key value rather than mutating the value on the original table).

**Comment by clason** [comment]
Sorry, but all these points were discussed and rejected:
* it is the explicit goal here for `vim.o` to return plain tables _just like `vim.g`_. Otherwise we end up at `vim.opt` (which we already have, so there's no point in changing `vim.o` to match). So this is very much against this proposal.
* the reason for using `vim.dict_` is that they are specific to "dict-like" (i.e., keyword-value style) tables; they are meaningless for list-like (and hence general) tables. The naming should reflect that.

**Comment by justinmk** [comment]
>  `vim.dict_set_key(table, key, value)` with `value = true` as default

Naming nit: do these functions need "key" in the name? `vim.dict_set` is unambiguous.

Also if we swap the position of value, we can accept variadic series of keys, so to set `vim.o.foo.key1.key2`:

```
vim.dict_set(vim.o.foo, val, 'key1', 'key2', ...)
```

**Comment by binyomen** [comment]
> it is the explicit goal here for vim.o to return plain tables just like vim.g. Otherwise we end up at vim.opt (which we already have, so there's no point in changing vim.o to match). So this is very much against this proposal.

Makes sense to me.

Would `dict_del_key` be setting the key to `nil`, or would there be additional functionality there? Just trying to gauge whether there'd be a difference between `vim.dict_del_key(tbl, key)` and `vim.dict_set_key(tbl, key, nil)`.

> Also if we swap the position of value, we can accept variadic series of keys, so to set `vim.o.foo.key1.key2`

It seems like it would be confusing to put the keys after the value for the function. Maybe there could be an alternative function `vim.dict_set_nested(tbl, {'key1', 'key2'}, val)` (very bad name, but that's the idea)?

**Comment by clason** [comment]
Table would be cleaner; we can dispatch on the type of argument (string vs. table) and do this without adding another function.

**Comment by binyomen** [comment]
I'd say if we use a table we should create another function. Any type can be used as keys for tables, including other tables. This is a fairly common pattern to give an object a unique key that won't overlap with any user-created ones. This wouldn't be the case for `vim.g` or `vim.o`, but if we want these functions to work generally for other tables as well I think we should allow tables to be keys.

**Comment by mg979** [comment]
Isn't it possible to have, for example

```lua
vim.o.rtp   -- returns a string
vim.o.rtp() -- returns a table
-- chainable metamethods
vim.o.rtp()
  :append(path) -- append a path to current runtimepath
  :prepend(path) -- prepend a path
```

And get rid of `vim.opt` obviously.

**Comment by lewis6991** [comment]
> Isn't it possible to have, for example
> 
> ```lua
> vim.o.rtp   -- returns a string
> vim.o.rtp() -- returns a table
> -- chainable metamethods
> vim.o.rtp()
>   :append(path) -- append a path to current runtimepath
>   :prepend(path) -- prepend a path
> ```
> 
> And get rid of `vim.opt` obviously.

No this is not possible

**Comment by mg979** [comment]
Why not? You could use the `__call` metamethod on the string to return the table, and implement additional metamethods on it (or directly on the string). I think it would be backwards compatible too.

**Comment by lewis6991** [comment]
You can't apply metamethods to strings.

**Comment by mg979** [comment]
How not, this works

```lua
local rtp = vim.o.rtp
local mt = getmetatable(rtp)

function mt.__call()
  return vim.opt.rtp:get()
end

print(rtp)
print(vim.inspect(rtp()))
```

**Comment by lewis6991** [comment]
That's changing the global table for all strings. We are not doing that!

**Comment by mg979** [comment]
Then as @justinmk said, `vim.o().rtp` can return a table and implement all metamethods there, so that

    vim.o().rtp
    vim.o().rtp:append(path)

behave like

    vim.opt.rtp:get()
    vim.opt.rtp:append(path)

Sorry for the noise.

**Comment by bew** [comment]
I don't understand the reasoning to expose _plain Lua tables_ at `vim.o` level, what's so appealing that you'd want to remove the `vim.opt` way?

What are the problems with the `vim.opt` way of doing things?
Why are you (@justinmk) against having a kind of (leaf) Option type (with meta methods, and/or helpers that manipulate Option values)
You say `we can do that (exposing structure) without adding special interfaces or opaque objects`
But IMO options seems like a sensible place for a dedicated Option type, with special interfaces to query/set them. They are such a central cornerstone of (neo)vim!

I really like @tjdevries' comment at https://github.com/neovim/neovim/issues/18246#issuecomment-1114020705
I also think `vim.o` interface is sub-optimal and could benefit from being removed to favor a more structured way to interact with options, with meta helpers to get/set/traverse values and its structure.

Maybe we can dumb it down a little, to try to remove some _magic_, but the interface has a good structure and helpers that shouldn't be dismissed.

IDEA: I'm also thinking we could offer a better set of option names by namespacing them like `gui.font`, `ui.statusline`, `wm.equalalways`, `formatopts.join_removes_comment`, ...
=> Would much improve discoverability (god knows how many times I discovered some options by mistake while navigating some unrelated options..) and better see the relations between options.
(What do you think about such an idea? If you say _not now_ I'd agree it might not be the time, maybe we can add this later?)

In https://github.com/neovim/neovim/issues/18246#issuecomment-1107931182 you say `A Lua table interface is useful _if it works with vanilla tables_.`
Then why not add more support for vanilla tables, to make them _feel like_ vanilla tables if you want, with good composition while offering a nice interface ?

In your proposal you mention:
> One difficulty is that we cannot put too much "metamagic" into these tables without losing the possibility to nicely print or otherwise work with them (hence the need for :get() in vim.opt).

For me printing them can easily be achieved with a `__tostring`, printing something like `<Option global "foobar">`.

Define `work with them`

> [...] The flip side is that the option keys cannot be manipulated individually (that would require adding an __index on the vim.o.listchars table)

Why not do that?

**Comment by clason** [comment]
You misunderstand the fundamental point: we do not want to handle table options (painfully) at the Lua level, we want to have options _natively_ (i.e., at the C core level) support list/dict options so it Just Works™️ from vimscript, Nvim API, Lua, ...

**Comment by feoh** [comment]
Hi.

Thanks for pointing this out. It's a good distant early warning.

I'm going to leave this bug around because we'll probably eventually need to do the work of translating all the calls, but we are a "leaf node" in the Neovim ecosystem.

We'll need ALL the plugins this project encompasses to be fixed first, then our configuration will follow suit.

**Comment by clason** [comment]
Anecdotally, the main usecase for `vim.opt` over `vim.o` is modifying `runtimepath` (followed by removing list options). So a dedicated `nvim_modify_runtimepath` API that allows removing or inserting (in specified positions) would already get us most of the way of being able to deprecate `vim.opt`, even without generic table options.

(Example use case: Nvim-treesitter allows setting custom directories for installing parsers and queries; these directories generally (but not in all cases, e.g. for testing) need to come before the "builtin" directories but after the "user" directories.)

**Comment by WieeRd** [comment]
How about breaking the baseline assumption that the options have to be handled as strings? I've always disliked how some of the internals of Vim seem to be designed around the syntax of Vimscript rather than the other way around.

What if `vim.o` and the table form of the option values become the primary, 'canonical' interface to accessing the options? Then `:set` becomes a wrapper that converts string values into a table. Would that require too many drastic changes to the codebase?

Edit: Never mind, I missed the very last line of the issue. Glad to know it is within consideration at least 👍 

> Even further, it could be useful to change the internal representation of these list- and dict-style options to actual tables and convert to string only when necessary, e.g. for access from Vimscript.

**Comment by clason** [comment]
@WieeRd See the last paragraph of the issue and https://github.com/neovim/neovim/issues/32139. This requires a _massive_ refactor of the legacy (Vim) C code and is unlikely to happen in the near future.

**Comment by superatomic** [comment]
A suggestion: Perhaps some/all of the existing methods for `vim.opt.*` could be moved to `vim.o`. Example: `vim.opt.spelloptions:append('camel')` would become `vim.o:append('spelloptions', 'camel')`.

**Comment by clason** [comment]
This has all been discussed before. TL;DR: No, this would just mean renaming `vim.opt` to `vim.o` and deprecating the current `vim.o`, which is the opposite of the agreed plan.

**Comment by superatomic** [comment]
It's not akin to deprecating the current `vim.o`. The methods would be on `vim.o` itself, not `vim.o.optionname`. This would be an addition to `vim.o`, which just extends its current functionality.

**Comment by clason** [comment]
Then this has been suggested already.

**Comment by drowning-cat** [comment]
A dumb alternative is to work with strings:

```lua
vim.o.something = vim.o_add("here,we,go,,", "again")
vim.o.something = vim.o_del(vim.o.something, "name")
```

Cons:

1. Can't handle option-specific exceptions.
2. Have to repeat the option.
